tests/System.Text.Formatting.Tests/Parsing/ParserTests.cs

```diff
@@ -0,0 +1,340 @@
+            int parsedValue;
+            int bytesConsumed;
+            bool result = InvariantParser.TryParse(UtfEncode(text), index, out parsedValue, out bytesConsumed);
+
+            Assert.Equal(expectSuccess, result);
+            Assert.Equal(expectedValue, parsedValue);
+            Assert.Equal(expectedBytesConsumed, bytesConsumed);
+        }
+
+        [Theory]
+        [InlineData("1728", true, 0, 1728, 4)]
+        [InlineData("blahblahh1751110", true, 9, 1751110, 7)]
+        [InlineData("987abcdefg", true, 0, 987, 3)]
+        [InlineData("-2147483648", true, 0, -2147483648, 11)]
+        [InlineData("Letthem-32984eatcake", true, 7, -32984, 6)]
+        [InlineData("I am 1", false, 0, 0, 0)]
+        public unsafe void ParseUtf8ByteStarToInt(string text, bool expectSuccess, int index, int expectedValue, int expectedBytesConsumed)
+        {
+            int parsedValue;
+            int bytesConsumed;
+
+            GCHandle arrayPinned = GCHandle.Alloc(UtfEncode(text), GCHandleType.Pinned);
+            byte* arrayPointer = (byte*)arrayPinned.AddrOfPinnedObject();
+
+            bool result = InvariantParser.TryParse(arrayPointer, index, 50, out parsedValue, out bytesConsumed);
+
+            Assert.Equal(expectSuccess, result);
+            Assert.Equal(expectedValue, parsedValue);
+            Assert.Equal(expectedBytesConsumed, bytesConsumed);
+        }
+        #endregion
+
+        #region short
+        [Theory]
+        [InlineData("1728", true, 0, 1728, 4)]
+        [InlineData("blahblahh17511", true, 9, 17511, 5)]
+        [InlineData("987abcdefg", true, 0, 987, 3)]
+        [InlineData("-32768", true, 0, -32768, 6)]
+        [InlineData("Letthem-32684eatcake", true, 7, -32684, 6)]
+        [InlineData("-A", false, 0, 0, 0)]
+        [InlineData("I am 1", false, 0, 0, 0)]
+        public void ParseUtf8ByteArrayToShort(string text, bool expectSuccess, int index, short expectedValue, int expectedBytesConsumed)
+        {
+            short parsedValue;
+            int bytesConsumed;
+            bool result = InvariantParser.TryParse(UtfEncode(text), index, out parsedValue, out bytesConsumed);
+
+            Assert.Equal(expectSuccess, result);
+            Assert.Equal(expectedValue, parsedValue);
+            Assert.Equal(expectedBytesConsumed, bytesConsumed);
+        }
+
+        [Theory]
+        [InlineData("1728", true, 0, 1728, 4)]
+        [InlineData("blahblahh17511", true, 9, 17511, 5)]
+        [InlineData("987abcdefg", true, 0, 987, 3)]
+        [InlineData("-32768", true, 0, -32768, 6)]
+        [InlineData("Letthem-32684eatcake", true, 7, -32684, 6)]
+        [InlineData("I am 1", false, 0, 0, 0)]
+        public unsafe void ParseUtf8ByteStarToShort(string text, bool expectSuccess, int index, short expectedValue, int expectedBytesConsumed)
+        {
+            short parsedValue;
+            int bytesConsumed;
+
+            GCHandle arrayPinned = GCHandle.Alloc(UtfEncode(text), GCHandleType.Pinned);
```

Comment:
You need to "unpin" (free) if you pin using GCHandles. Though since you only need this pinned in this lexical scope, you should use "fixed" statement to pin, so you don't have to worry about freeing.